lib/Sema/SemaBounds.cpp

```diff
@@ -3285,6 +3288,318 @@ namespace {
+  ComparisonSet Union(ComparisonSet& S1, ComparisonSet& S2) {
+    if (S1.size() == 0)
+      return S2;
+    if (S2.size() == 0)
+      return S1;
+    ComparisonSet Result(S1);
+    Result.insert(S2.begin(), S2.end());
+    return Result;
+  }
+
+  // Given two sets S1 and S2, this function returns true if the two sets
+  // are not equal.
+  // Equality means that their sizes are the same, and every member of S1 is found in
+  // S2 and vice versa.
+  bool Differ(ComparisonSet& S1, ComparisonSet& S2) {
+    if (S1.size() != S2.size())
+      return true;
+    if (S1.size() == 0 && S2.size() == 0)
+      return false;
+    if (S1.size() > S2.size())
+      for (auto E : S1)
+        if (S2.find(E) == S2.end())
+          return true;
+    if (S1.size() < S2.size())
+      for (auto E : S2)
+        if (S1.find(E) == S1.end())
+          return true;
+    return false;
+  }
+
+  // Given two sets S1 and S2, the return value is the intersection of these sets.
+  ComparisonSet Intersect(ComparisonSet& S1, ComparisonSet& S2) {
+    if (S1.size() == 0)
+      return S1;
+    if (S2.size() == 0)
+      return S2;
+    ComparisonSet Result;
+
+    for (auto E1 : S1)
+      if (S2.find(E1) != S2.end())
+        Result.insert(E1);
+    return Result;
+  }
+
+  // This function returns true if variable V is used in the comparison I.
+  bool ContainsVariable(Comparison& I, const VarDecl *V) {
+    ExprSet Exprs;
+    CollectExpressions(I.first, Exprs);
+    CollectExpressions(I.second, Exprs);
+    for (auto InnerExpr : Exprs)
+      if (const DeclRefExpr *DRE = dyn_cast<DeclRefExpr>(InnerExpr))
+        if (const VarDecl *VD = dyn_cast<VarDecl>(DRE->getDecl()))
+          if (VD == V)
+            return true;
+    return false;
+  }
+
+  // This function looks recursively for comparisons in statement `S`.
+  // Then, it records them as elements of type Comparison in `ISet`.
+  // Each Comparison is a pair in form of (Expr1, Expr2) where both
+  // Expr1 and Expr2 are of type Expr*.
+  // Each pair corresponds to a comparison expression of one of the following forms:
+  // - Expr1 < Expr2
+  // - Expr1 <= Expr2
+  // - Expr2 > Expr1
+  // - Expr2 >= Expr1
+  void ExtractComparisons(const Stmt *St, ComparisonSet &ISet) {
+    if (const BinaryOperator *BO = dyn_cast<BinaryOperator>(St)) {
+      switch (BO->getOpcode()) {
+        case BinaryOperatorKind::BO_LOr:
+          return;
+        case BinaryOperatorKind::BO_LAnd:
+          return;
+        case BinaryOperatorKind::BO_LE:
+        case BinaryOperatorKind::BO_LT:
+          ISet.insert(Comparison(BO->getLHS(), BO->getRHS()));
+          break;
+        case BinaryOperatorKind::BO_GE:
+        case BinaryOperatorKind::BO_GT:
+          ISet.insert(Comparison(BO->getRHS(), BO->getLHS()));
+          break;
+        default:
+          break;
+      }
+    } else
+      return;
+    for (auto I = St->child_begin(); I != St->child_end(); ++I)
```

Comment:
This logic is wrong - you shouldn't recurse.     There's no guarantee that comparison produces a true or false value consumed by the if-statement.  For example, consider `if (f(a < b))`.